### PR TITLE
Flight: Add raw and comp GPS messages to MSP protocol

### PR DESCRIPTION
**Pull request by jihlein**
Bench tested at various home and ownship positions and ranges.  MWOSD display seems to be correct.  

Need someone to check my math.  Also think about what math simplifications might be made given the display resolution.

[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/571)

### CODE REVIEW

**Comment by dustin** [comment]
Awesome.  I might see if I can get this going on my naze.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 336 [r1]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EAnfuwxQ6r3N9hGIG:-K9EAnfvtkqllrNmSxfw:444867413) ([raw file](https://github.com/d-ronin/dronin/blob/0408d06e48f8a16775c85bfe8efcb4774a2ef5b2/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L336)):_
These two lines have odd horizontal spacing.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/571#-:-K9EAt51zW52JMew9hII:-302977716)_

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 1]
```diff
@@ -322,12 +329,72 @@ static void msp_send_ident(struct msp_bridge *m)
...
+	
+	int32_t deltaLon = (homeData.Longitude - gpsData.Longitude);  // degrees * 1e7
+	int32_t deltaLat = (homeData.Latitude  - gpsData.Latitude );  // degrees * 1e7
+	
+	float deltaY = (float)deltaLon * WGS84_RADIUS_EARTH_KM * DEG2RAD;  // KM * 1e7
+	float deltaX = (float)deltaLat * WGS84_RADIUS_EARTH_KM * DEG2RAD;  // KM * 1e7
+	
+	deltaY *= cosf((float)homeData.Latitude * 1e-7f * (float)DEG2RAD);  // Latitude compression correction
+	
+	data.comp_gps.distance_to_home  = (uint16_t)(sqrt(deltaX * deltaX + deltaY * deltaY) * 1e-4);  // meters
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: sqrtf okay here rather than double precision sqrt?

**Comment by jihlein** [comment]
Good catch, replaced with sqrtf.  Also should have mentioned my testing has been with an AQ32 board.

**Comment by mlyle** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 263 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EUZxHkup96Cx7MN5e:-K9EUZxHkup96Cx7MN5f:-289064107) ([raw file](https://github.com/d-ronin/dronin/blob/c907e16adc63b4ca58ef8984a0d52403bcb101d1/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L263)):_
These lines are space-indented too

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 347 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EUi6VhijmAb3rChXS:-K9EUi6VhijmAb3rChXT:1916723154) ([raw file](https://github.com/d-ronin/dronin/blob/c907e16adc63b4ca58ef8984a0d52403bcb101d1/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L347)):_
If there's not GPS, what should we do?  Probably not what we do below, e.g. return 0's..  Should we just return and decline to answer?

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 373 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EVV8Yt85PxWUt0vKR:-K9EVV8Yt85PxWUt0vKS:-767890270) ([raw file](https://github.com/d-ronin/dronin/blob/c907e16adc63b4ca58ef8984a0d52403bcb101d1/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L373)):_
Even more important here-- if GPS is missing, return the direction to homeloc from 0,0?  Probably should just return.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/571)_

**Comment by jihlein** [comment]
Line 265 covers this, if there is no GPS fix, all the GPS information is suppressed from the display.  We could still return on GPS not found too.

I'm not following the spacing comments, they appear to be the same as the other instances in the file.

**Comment by dustin** [comment]
It's perfectly fine to not return a requested value AFAIK.

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/571#-:-K9EYvLgpEK05hleDSSk:1682648746)_

**Comment by mlyle** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 336 [r1]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EAnfuwxQ6r3N9hGIG:-K9EZiiks3P7216PIjM-:-1333159692) ([raw file](https://github.com/d-ronin/dronin/blob/0408d06e48f8a16775c85bfe8efcb4774a2ef5b2/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L336)):_
These ones look fine to me actually

---

_[flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c, line 263 [r2]](https://reviewable.io:443/reviews/d-ronin/dronin/571#-K9EZpjyPCg4CQankp30:-K9EZpjyPCg4CQankp31:175620115) ([raw file](https://github.com/d-ronin/dronin/blob/c907e16adc63b4ca58ef8984a0d52403bcb101d1/flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c#L263)):_
These lines are one tab then 16 spaces to try and align; should be two tabs. (and give up on maintaining alignment).

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/571)_

**Comment by mlyle** [comment]
Yah-- whether or not they're not going to use the info, it's better to not send back garbage information anyways.

**Comment by tracernz** [comment]
[MSP defines an error response](http://www.multiwii.com/forum/viewtopic.php?f=8&t=1516) in case you decline to answer:

```
$M|[0][code][checksum]
1 octet '$'
1 octet 'M'
1 octet '|'
1 octet 0
1 octet [unknown code]
1 octet [checksum]
```

Cleanflight does something different... https://github.com/cleanflight/cleanflight/blob/master/src/main/io/serial_msp.c#L207
And MWOSD doesn't care unless it's a > https://github.com/ShikOfTheRa/scarab-osd/blob/master/MW_OSD/Serial.ino#L824

**Comment by jihlein** [comment]
So the issue becomes when do we display raw GPS data, and when do we have good enough GPS data to display meaningful computed GPS data (distance to home and direction to home).

We have the following GPS states: NoGPS ,NoFix, Fix2D, Fix3D, Diff3D

We could display raw GPS at Fix2D or greater.  Home position is set at Fix3D, so we could start displaying computed GPS data then.  If drops below Fix3D, do we continue to display the computed data?

This is further complicated by the MWOSD software, which seems to display both the raw and computed GPS data once GPS is set in the sensor word.  I need to research this a little further and see if there is a way to independently select raw and computed displays.

**Comment by mlyle** [comment]
@jihlein yes, appropriately masking the data would be even better.  I don't know the answer.

**Comment by jihlein** [comment]
@mlyle, a possible answer is to modify the MWOSD code slightly, to allow for individually masking the raw gps vs the comp gps data.  We could turn on/off the MWOSD mods with a Dronin specific #define.

My thought is to display raw gps if fix >= 2d.  The comp gps won't be displayed until a home position is set, and once displayed, will be displayed as long as fix >= 2d.  Fix < 2d will cause all GPS data to disappear.

It's only a few lines of change to the MWOSD code, so I'm going to try this later today.  It's one more thing to have to load/configure, but at least this aligns more with the GPS paradigm we use.  I don't know how *flight code sets this up, not sure it really matters.

Plus, it's one less excuse to not make a custom character set and have the MWOSD correctly display Dronin flight modes.  But that's for another day/discussion......

**Comment by mlyle** [comment]
@jihlein There's already #ifdef TAULABS in MWOSD.  Just-- in my opinion-- as much as possible of this stuff should work with the plain MWOSD-as-distributed.  So, whatever you send should have the best possible "fallback" behavior on the existing MWOSD code.

**Comment by jihlein** [comment]
Well I reworked the logic to send zeros when gps and/or home position is invalid.

It's not a 100% solution though, the home distance and direction display as 0 when they really should not be displayed.  Standard MWOSD code has raw and computed GPS display tied together.

A 99% solution is new MWOSD code.  You can find the code I'm using at:
https://github.com/jihlein/scarab-osd/tree/dronin

It's not a 100% solution because there is another firmware package that must be loaded into the OSD hardware.

If I'm missing something in the MWOSD firmware, I'm not seeing it.

File: flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c [hunk 2]
```diff
@@ -322,12 +329,99 @@ static void msp_send_ident(struct msp_bridge *m)
...
+		data.comp_gps.distance_to_home    = 0;
+		data.comp_gps.direction_to_home   = 0;
+		data.comp_gps.home_position_valid = 0;  // Home distance and direction will not display on OSD
+	}
+	else
+	{
+		GPSPositionGet(&gps_data);
+		HomeLocationGet(&home_data);
+		
+		if((gps_data.Status < GPSPOSITION_STATUS_FIX2D) || (home_data.Set == FALSE))
```
Comment (tracernz) [inline flight/Modules/UAVOMSPBridge/UAVOMSPBridge.c]: HOMELOCATION_SET_FALSE :P

**Comment by mlyle** [comment]
LGTM.

**Pull request merged**